content/6_Advanced/String_Search.mdx

```diff
@@ -188,15 +186,58 @@ int main() {
 			r = i + z[i];
 			z[i]++;
 		}
+	}
 
-		if (z[i] + i == s.size()) { cout << i << ' '; }
+	return z;
+}
+
+int main() {
+	string s;
+	cin >> s;
+	const auto z = z_function(s);
+
+	for (int i = 0; i < s.size(); i++) {
+		if (z[i] + i == s.size()) { cout << i << " "; }
 	}
 
 	cout << s.size() << endl;
 }
 ```
 
 </CPPSection>
+<PySection>
+
+```py
+from typing import List
+
+
+def z_function(s: str) -> List[int]:
+	n = len(s)
+	z = [n]
+	l, r = 0, 0
+	for i in range(1, n):
+		zi = min(r - i, z[i - l]) if i < r else 0
+		while i + zi < n and s[i + zi] == s[zi]:
+			zi += 1
+		if i + zi > r:
+			l, r = i, i + zi
+		z.append(zi)
+	return z
+
+
+def main():
```

Comment:
is this needed

Reply:
like putting the code inside a function instead of the file directly

Reply:
I prefer `main` to putting stuff in global scope because it has a local scope, thus avoiding clashes between variables. However I think for the brevity purposes we can get rid of `main`. Updated.